core/src/main/java/com/sequenceiq/cloudbreak/core/cluster/ClusterManagerMemoryAdjuster.java

```diff
@@ -0,0 +1,133 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClusterManagerMemoryAdjuster.class);
+
+    private static final double MAX_CM_MEMORY_USAGE_FACTOR = 0.25;
+
+    private static final Memory FOUR_GIGABYTES = Memory.ofGigaBytes(4);
+
+    private static final Memory THIRTY_TWO_GIGABYTES = Memory.ofGigaBytes(32);
+
+    private static final int ONE_HUNDRED = 100;
+
+    private static final int MIN_CM_MEMORY_NODE_COUNT = ONE_HUNDRED;
+
+    private static final int MAX_CM_MEMORY_NODE_COUNT = 800;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private ClusterApiConnectors clusterApiConnectors;
+
+    @Inject
+    private CloudbreakEventService cloudbreakEventService;
+
+    public void adjustMemory(StackDto stackDto, int nodeCount) {
+        if (!StackType.WORKLOAD.equals(stackDto.getType())) {
+            LOGGER.info("Memory adjustment is skipped. It is only supported for data hub clusters.");
+            return;
+        }
+        try {
+            GatewayConfig gateway = gatewayConfigService.getPrimaryGatewayConfig(stackDto);
+            Memory currentMemory = getClusterManagerMemory(gateway);
+            Memory requiredMemory = calcRequiredClusterManagerMemory(nodeCount);
+            if (requiredMemory.getValueInBytes() > currentMemory.getValueInBytes()) {
+                Memory availableMemory = getAvailableMemory(gateway);
+                Memory maxAllowedMemory = Memory.ofGigaBytes((int) (availableMemory.getValueInGigaBytes() * MAX_CM_MEMORY_USAGE_FACTOR));
+                if (requiredMemory.getValueInBytes() > maxAllowedMemory.getValueInBytes()) {
+                    sendWarningToUser(stackDto, requiredMemory, availableMemory);
+                    requiredMemory = maxAllowedMemory;
+                }
+
+                if (requiredMemory.getValueInBytes() > currentMemory.getValueInBytes()) {
+                    LOGGER.info("Updating cluster manager memory. Current value: {}, Target value: {}, Available memory: {}",
+                            currentMemory, requiredMemory, availableMemory);
+                    updateClusterManagerConfig(gateway, requiredMemory);
+                    restartClusterManager(gateway, stackDto);
+                    witForClusterManagerToBecomeAvailable(stackDto);
+                }
+            } else {
+                LOGGER.info("Cluster manager's configured memory {}GB is enough for {} nodes.", currentMemory, nodeCount);
+            }
+        } catch (Exception e) {
+            LOGGER.error("Error happened while tried to configure cluster manager memory.", e);
+        }
+    }
+
+    private Memory getClusterManagerMemory(GatewayConfig gatewayConfig) throws CloudbreakOrchestratorFailedException {
+        return hostOrchestrator.getClusterManagerMemory(gatewayConfig)
+                .orElseThrow(() -> new RuntimeException("Couldn't get cloudera manager memory."));
```

Comment:
Can we use CloudbreakRuntimeException please?

Reply:
Yes, sure.